state/upgrades.go

```diff
@@ -668,3 +671,38 @@ func AddControllerLogPruneSettings(st *State) error {
 	}
 	return nil
 }
+
+// AddStatusHistoryPruneSettings adds the model settings
+// to control log pruning if they are missing.
+func AddStatusHistoryPruneSettings(st *State) error {
+	coll, closer := st.getRawCollection(settingsC)
+	defer closer()
+
+	settingsChanged := false
+
+	iter := coll.Find(bson.M{"_id": bson.M{"$regex": "^.*:e$"}}).Iter()
```

Comment:
This feels like it'd be a bit too easy to stomp on unrelated settings docs. i.e., we might introduce some settings key with a prefix like model-uuid:foo:e, where e is some key that has nothing to do with environments.

Can we please use AllModels to iterate models, and then look up the settings docs for those? I think you could then refactor and share some code between AddControllerLogPruneSettings and AddStatusHistoryPruneSettings: call some common function with a collection, doc ID, and a function with the signature:

    func(*settingsDoc) (changed bool)